lightning/src/chain/keysinterface.rs

```diff
@@ -1133,9 +1176,108 @@ impl KeysInterface for KeysManager {
 		InMemorySigner::read(&mut io::Cursor::new(reader), self.get_node_secret())
 	}
 
-	fn sign_invoice(&self, hrp_bytes: &[u8], invoice_data: &[u5]) -> Result<RecoverableSignature, ()> {
+	fn sign_invoice(&self, hrp_bytes: &[u8], invoice_data: &[u5], invoice_type: Invoice) -> Result<RecoverableSignature, ()> {
 		let preimage = construct_invoice_preimage(&hrp_bytes, &invoice_data);
-		Ok(self.secp_ctx.sign_recoverable(&hash_to_message!(&Sha256::hash(&preimage)), &self.get_node_secret()))
+		let secret = match invoice_type {
+			Invoice::Bolt11 => self.get_node_secret(),
+			Invoice::Phantom => return Err(()),
```

Comment:
Hmm, do we still want to support users that want a custom `KeysInterface` that could potentially wrap a `PhantomKeysManager`? Seems nice to keep flexibility there

Reply:
True, could be useful. Not sure what we'd see in practice. Could always make a sub-trait for phantom functionality, too.

Reply:
> Could always make a sub-trait for phantom functionality, too.

I guess it's not the best option to make a new trait containing both `sign_phantom_invoice` and `get_phantom_secret`, since `ChannelManager` currently calls out to `get_phantom_secret` and therefore `KeysManager` on its own wouldn't be viable for a time.

We could make a trait containing only `sign_phantom_invoice`, though `KeysInterface` would still be polluted with some phantom-ness until we implement getting rid of both `get_*_secret` methods. Not sure what the thinking is for timeline on this refactor.

Reply:
Yea, I'm not a huge fan of a second trait here because we still end up with phantom *stuff* in the `KeysInterface` so that `ChannelManager` can decrypt onions to the phantom key. We could add a whole second interface that `ChannelManager` uses to get the phantom keys, with associated parameterizations, etc, or we could just have two of the methods take a `invocie` parameter. Its maybe arguably "cleaner" to have a separate interface, but its so much additional overhead for such a marginal difference I don't really get why we'd go to the trouble.

As for the timeline on the removal of `get_*_secret`, it doesn't change this too much - when we drop it we'll still have a `decrypt_onion` method that will now have to take the `Invoice` parameter, instead of the secret-key-fetching method.

Reply:
Yeah, all fair points regarding the trait. I guess it feels a little odd to have `sign_invoice` where we'd want to end up but not do the same for decrypting hops. Is it a large change to add `decrypt_onion` even if we don't immediately hide `get_node_secret`?

Reply:
> Is it a large change to add decrypt_onion even if we don't immediately hide get_node_secret?

Hmm, likely not - should just pass in the ephemeral PK, the `Invoice` mode enum, and then have it return the result of `gen_rho_mu_from_shared_secret`. That code is already being refactored in this PR. As a more trivial alternative, we could of course change the API of `get_node_secret` to take the `Invoice` mode enum.

Reply:
Either way sounds fine by me.

Reply:
I left the `Invoice` enum as-named since we should be getting rid of `get_node_secret` near-term